### ISSUE TITLE
Consider adding backward-compatibility to the kernel .dill files

### ISSUE THREAD

**Issue by aam** [dart-lang/sdk#30192]
To add Flutter flavor to https://github.com/dart-lang/sdk/issues/30188, Flutter app kernel .dill file is produced with one version of dart sdk on a workstation, consumed by another running on a device. Would it be possible to have some guarantees regarding backward compatibility for .dill files, so whatever is produced on the workstation can be consumed on the device?

Reason there are two is that:
1) Flutter tools running on the workstation have been using dart sdk to run pub and analyzer. This will change with adding kernel support, where .dill file will be produced when running flutter app in debug mode. That [dart sdk](https://github.com/flutter/flutter/blob/master/bin/internal/dart-sdk.version) moves slow. 
2) Flutter engine running on the device has another dart sdk 'built-in', that is [built](https://github.com/flutter/engine/blob/master/DEPS#L33) and included into engine artifacts. Engine artifacts are downloaded from the cloud by Flutter tools separately from dart sdk(https://github.com/flutter/flutter/blob/master/bin/internal/engine.version).

So to summarize, for Flutter tools dart-sdk.version moves slower, engine.version moves fast, and if there is significant gap between them, kernel files produced on the workstation won't be loaded on the device.

+a-siva
+sigmundch

**Comment by aam** [comment]
@Hixie , I remember someone mentioned a reason why Flutter tools track dart sdk at slower pace than Flutter engine, but I don't remember what it was. If there is no good reason, we might be keeping dart-[sdk.version](https://github.com/flutter/flutter/blob/master/bin/internal/dart-sdk.version) in sync with [engine DEPS](https://github.com/flutter/engine/blob/master/DEPS#L33) file reference whenever we advance [engine.version](https://github.com/flutter/flutter/blob/master/bin/internal/engine.version).

**Comment by Hixie** [comment]
I'm not aware of any reason. Keeping them exactly in sync would be ideal. cc @pq

**Comment by pq** [comment]
I'm not sure what the rationale for the lag is; if there is one, maybe it has to do w/ staying in sync with the internal SDK?  @tvolkert might have more to say about that.

All of that being said, no skew would be awesome for all the obvious reasons! 👍

**Comment by tvolkert** [comment]
There's no inherent reason.

If we can make [dart-sdk.version](https://github.com/flutter/flutter/blob/master/bin/internal/dart-sdk.version) reference an exact SHA1 hash, then yes, we could just keep it in sync with the SDK version that's in the DEPS file of the matching engine version.

Or potentially even better, see https://github.com/flutter/flutter/issues/7437#issuecomment-311474355 -- we could build a standalone embedder in the engine against the version in our DEPS file and ship that with the engine so that we don't even have to maintain a separate dart-sdk.version file in flutter/flutter...

**Comment by tvolkert** [comment]
One potential gotcha to be aware of is that if we start using new Dart SDK APIs in flutter_tools before those APIs are available inside Google, then we'll be blocked on rolling Flutter into Google, which we need to avoid.

**Comment by aam** [comment]
@a-siva since we are considering relying on kernel files backward-compatibility, I just ran a test where flutter tools runs on dart sdk 1.25.0-dev.7.0, flutter engine dart is on a83a69ff3eff845980308779cadbdb80cc03a603 (Jul 18). In this setup main.dart.dill produced by flutter tools fails to run on flutter engine:
```
[+1920 ms] E/flutter (27161): [ERROR:lib/tonic/logging/dart_error.cc(16)] Unhandled exception:
[        ] E/flutter (27161): type 'int' is not a subtype of type 'StackTrace' of 'stackTrace' where
[        ] E/flutter (27161):   int is from dart:core
[        ] E/flutter (27161):   StackTrace is from dart:core
[        ] E/flutter (27161): 
[        ] E/flutter (27161): #0      _setAsyncThreadStackTrace (dart:async)
[        ] E/flutter (27161): #1      SystemChrome.setApplicationSwitcherDescription (package:flutter/src/services/system_chrome.dart)
[        ] E/flutter (27161): #2      new Future.microtask.<anonymous closure> (flutter_patched_sdk/lib/async/future.dart:200:37)
[        ] E/flutter (27161): #3      _microtaskLoop (flutter_patched_sdk/lib/async/schedule_microtask.dart:41:21)
[        ] E/flutter (27161): #4      _startMicrotaskLoop (flutter_patched_sdk/lib/async/schedule_microtask.dart:50:5)
[   +4 ms] E/flutter (27161): [ERROR:lib/tonic/logging/dart_error.cc(16)] Unhandled exception:
[        ] E/flutter (27161): Invalid argument(s)
[        ] E/flutter (27161): #0      _setAsyncThreadStackTrace (dart:async)
[        ] E/flutter (27161): #1      CachingAssetBundle._fetchString (package:flutter/src/services/asset_bundle.dart)
[        ] E/flutter (27161): #2      new Future.microtask.<anonymous closure> (flutter_patched_sdk/lib/async/future.dart:200:37)
[        ] E/flutter (27161): #3      _microtaskLoop (flutter_patched_sdk/lib/async/schedule_microtask.dart:41:21)
[        ] E/flutter (27161): #4      _startMicrotaskLoop (flutter_patched_sdk/lib/async/schedule_microtask.dart:50:5)
[        ] E/flutter (27161): [ERROR:lib/tonic/logging/dart_error.cc(16)] Unhandled exception:
[        ] E/flutter (27161): type 'int' is not a subtype of type 'StackTrace' of 'stackTrace' where
[        ] E/flutter (27161):   int is from dart:core
[        ] E/flutter (27161):   StackTrace is from dart:core
[        ] E/flutter (27161): 
[        ] E/flutter (27161): #0      _setAsyncThreadStackTrace (dart:async)
[        ] E/flutter (27161): #1      OptionalMethodChannel.invokeMethod (package:flutter/src/services/platform_channel.dart)
[        ] E/flutter (27161): #2      new Future.microtask.<anonymous closure> (flutter_patched_sdk/lib/async/future.dart:200:37)
[        ] E/flutter (27161): #3      _microtaskLoop (flutter_patched_sdk/lib/async/schedule_microtask.dart:41:21)
[        ] E/flutter (27161): #4      _startMicrotaskLoop (flutter_patched_sdk/lib/async/schedule_microtask.dart:50:5)
```

**Comment by aam** [comment]
After @a-siva and me looked closer at the stacktrace above, it became clear the the root cause of the failures is not kernel file format per se, rather the fact that kernel file didn't have changes needed to get causal_async_stacks flag working in VM. These changes were introduced in https://codereview.chromium.org/2982943002/ and that cl didn't make it into 1.25.0-dev.7.0.

**Comment by thosakwe** [comment]
Is there any update on this?
